### DISCUSSION TITLE
Problem with the "net_lwip_webserver (RNDIS) example"

### DISCUSSION THREAD

**Discussion by Wini-Buh** [hathach/tinyusb#853] [Q&A]
Hi

I'm tying to get the net_lwip_webserver (RNDIS) example working. Unfortunately there is problem with the notification end point (0x81). Everything I tried did not helped to solve the problem. It seems that the end point does not send the RNDIS "report data" (this is send here https://github.com/hathach/tinyusb/blob/4e2684f23aacfd9a08046c2a60da50bd1af09a27/lib/networking/rndis_reports.c#L81 (at least I could not see any communication in the Wireshark on that end point).

At the moment I also work on the on porting the tinyUSB to the CCRX toolchain from Renesas (see #785 ... I will also do a draft PR this week with all my changes for a deeper discussion about an inclusion into the tinyUSB project) 

I know it is difficult to ask for some support in this stage, but I don't have any idea how to solve the problem. In the attached debugging output of the tinyUSB, you see an assert is triggered.

Thank you very much for an answer.
[tinyusb_rndis.log](https://github.com/hathach/tinyusb/files/6551998/tinyusb_rndis.log)

**Comment by hathach** [comment]
@Wini-Buh thank you very much for your effort, the rndis is considered as advanced examples. Not all the existing ports actually run fine with it https://github.com/hathach/tinyusb/issues/289 . I have an list of running ports, though I am kept busy with other works to actually verify them all. Anyway, don't worry too much of about that particular example. The initial porting PR does not requires all example to run. Bugs are expected. I think you only need to get the msc_cdc and or hid_composite running, that would be just good enough. The rest can be addressed via follow up PRs.

**Reply by Wini-Buh** [reply]
Please let me explain briefly. I am working on a project that wants to integrate RNDIS over USB. Before I found tinyUSB, I tried the USB stack from Reneas. Unfortunately, I was not able to extend it to include RNDIS functionality (I got the enumeration to work, but had some issues in the code that I could not overcome).

I will try your suggestion to get the msc_cdc or hid_composite running. Do have a suggestion which one is easier to do ?

On the other side to get a progress with the RNDIS code (it seems there is one implementation around the web, because I used more or less the same source code base, when I tried to get RNDIS working with the Renesas USB stack), did you had a look into the debug output I attached to this question ? Could you give me some hint where the problem could be ?

Thank a lot for you help.

**Reply by hathach** [reply]
There is an assert within your log, the location suggest that the endpoint is already busy, which mean the previous transfer is not complete. This can be seen on the log, the endpoint 81 has been queued with 8 bytes, but there is no complete event on that event yet, until the 2nd attempt, which assert the error. Could you check to see if the interrupt endpoint on renesas work with 8 bytes ?

both msc_cdc and hid_composite should be running, one is bulk only, and other is interrupt endpoint. I normally will go with cdc_msc first with each new port.

```
USBD Setup Received 21 00 00 00 00 00 00 18 
  NET control request
  Queue EP 00 with 24 bytes ... OK
USBD Xfer Complete on EP 00 with 24 bytes
  NET control complete
  Queue EP 81 with 8 bytes ... OK
  Queue EP 80 with 0 bytes ... OK
USBD Xfer Complete on EP 80 with 0 bytes

USBD Setup Received A1 01 00 00 00 00 04 01 
  NET control request
  Queue EP 80 with 52 bytes ... OK
USBD Xfer Complete on EP 80 with 52 bytes
  NET control complete
  Queue EP 00 with 0 bytes ... OK
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 21 00 00 00 00 00 00 20 
  NET control request
  Queue EP 00 with 32 bytes ... OK
USBD Xfer Complete on EP 00 with 32 bytes
  NET control complete
  Queue EP 81 with 8 bytes ... usbd_edpt_xfer 1245: assert failed
  Queue EP 80 with 0 bytes ... OK
USBD Xfer Complete on EP 80 with 0 bytes
````

**Reply by Wini-Buh** [reply]
> There is an assert within your log, the location suggest that the endpoint is already busy, which mean the previous transfer is not complete. This can be seen on the log, the endpoint 81 has been queued with 8 bytes, but there is no complete event on that event yet, until the 2nd attempt, which assert the error. Could you check to see if the interrupt endpoint on renesas work with 8 bytes ?

That is exactly what I have found out today. I don't know exactly if the problem occurs while the data is written into the FIFO. Fact is there is no interrupt at all on endpoint 81, maybe there is a wrong configuration on that endpoint ? I will check that.
If I understood everything correctly when studying the HW manual, 8 bytes for USB interrupt transfer should not be a problem.

Ok, I will also try to go with the cdc_msc in the next days.

**Reply by Wini-Buh** [reply]
I checked today if something is wrong while copy the data to the FIFO. All seems to fine if have look at the several state flags in the USB HW registers. The bits signal a correct switch from the FIFO buffer on the CPU side to the SIE (Serial Interface Engine) side if all 8 bytes are written into the FIFO, but the transfer is not started.

Do you have five minutes to check the attached small Wireshark capture file ? Maybe you will see something that could be a problem (of course except the missing communication over endpoint 81).

[tinyusb_net_lwip_webserver.zip](https://github.com/hathach/tinyusb/files/6580538/tinyusb_net_lwip_webserver.zip)

**Reply by hathach** [reply]
@Wini-Buh you really need to get the other 2 examples running first, it would be much easier afterwards.

**Reply by kkitayam** [reply]
I checked the net_lwip_webserver example in the current master's HEAD using GR_CITRUS(RX63N) on windows 10. It seems to be working fine. chrome can open http://192.168.7.1/ . I guess there is some kind of endianness problem in the USB stack.
[gr_citrus.zip](https://github.com/hathach/tinyusb/files/6586177/gr_citrus.zip) is the Wireshark capture file regarding net_lwip_webserver.

**Reply by Wini-Buh** [reply]
@kkitayam Thanks a lot, I will take a look at the Wireshark capture and also try to get the 2 other examples working like @hathach did mentioned

**Reply by hathach** [reply]
@Wini-Buh are you testing on ubuntu, if yes, you may need to actively switch the network connection from your eithernet/wifi to the USB interface of the example.

**Reply by Wini-Buh** [reply]
No, I'm testing on Solus. And yes I know the fact with the network interface (I was running an other RNDIS example build for a STM32 device (with the Keil toolchain) and with this example I could see the new available network interface and access it).

I did porting the msc_cdc and hid_composite examples. Both are running now with my "extended build environment" (with CCRX in big and little endian configuration and with the GCC in little endian configuration). I had to correct an other issue in the USB driver for the Renesas Rx device (I will report that soon here). After correcting it everything worked (did also a quick check with it on the RNDIS port I'm working on, but without any success... same problem as before). Maybe I should try to get the RNDIS example getting running with the GCC to see where the problem could be ?

BTW: How do you test the hid_composite example ? If I connect it to my Linux host I'm losing control over the mouse (the mouse pointer is moving to the lower right corner of the screen).

**Reply by hathach** [reply]
you need to check your bsp board_button_read() implementation, it return 1 when button is pressed, it is likely that your bsp return the pin level (where high is inactive, gnd is active). The device will move mouse toward the bottom right corner and/or sending other keyboard, lower volume.

**Reply by Wini-Buh** [reply]
After I got the msc and hid examples working (like @hathach suggested), I tried to get the net_lwip_webserver example working again. This time I used the current master's HEAD and took the example code to integrate it into my demo project. Then I created two different projects (both with GCC, so we don't have any porting problem here) one uses FreeRTOS and the other uses the super loop found in the example code.  Both projects have the debug output enabled, where the output of an assert is visible (like below). Both projects produce the problem around endpoint 81.

```
USBD Setup Received 21 00 00 00 00 00 20 00 
  NET control request
  Queue EP 00 with 32 bytes ... OK
USBD Xfer Complete on EP 00 with 32 bytes
  NET control complete
  Queue EP 81 with 8 bytes ... usbd_edpt_xfer 1249: assert failed
  Queue EP 80 with 0 bytes ... OK
USBD Xfer Complete on EP 80 with 0 bytes
```

Curiously, the Ethernet interface is created on the Linux machine despite the error (I don't know if every thing is correct at that point, because at the moment I only check if the enumeration on the USB is correct, without the correct handling of the lwip stack).

@kkitayam Could you please run the net_lwip_webserver example again but with enabled debug messages and check if you see this assert output ?

**Reply by kkitayam** [reply]
> @kkitayam Could you please run the net_lwip_webserver example again but with enabled debug messages and check if you see this assert output ?

@Wini-Buh 
Please find attached [gr_citrus_rtt.zip](https://github.com/hathach/tinyusb/files/6631093/gr_citrus_rtt.zip). The assertion message is not found in the log.

Note: I built and logged the following with Windows 10.
1. Comment out `#include <stdalign.h>` on `rndis_reposrt.s` L.30 because of OPTLIB has no file.
2. `make BOARD=gr_citrus RX_NEWLIB=0 LOG=2 LOGGER=rtt flash`
3. Launch RTT viewer

**Reply by kkitayam** [reply]
Unfortunately, I only have a windows machine. I cannot confirm that the example will work correctly on a Linux machine.

**Reply by Wini-Buh** [reply]
I also did a quick test with an old Windows 7 and  the assert is not triggered. The enumeration is done successfully and all communication over endpoint 81 is performed correctly.

The enumeration with a Linux host seems a bit different (at least what I see on the Wireshark captures). But I'm not the USB specialist (that maybe I should be ? ). Do you have any idea how to find the reason of the problem ? It must be something around endpoint 81, which is used for interrupt transfers.

**Reply by kkitayam** [reply]
>   Queue EP 81 with 8 bytes ... usbd_edpt_xfer 1249: assert failed

I think the assertion message indicates that the next transfer has requested by application layer, even though the previous transfer has not been completed. It may be a bug regarding exclusive control operations in `dcd_usba.c`. I had thought that disabling USB interrupts would be sufficient.

**Reply by kkitayam** [reply]
@Wini-Buh 
Could you share the full log getting by Linux host? I would like to check transfers around the assertion message.

Interrupt IN transfers are invoked at [L.307](https://github.com/hathach/tinyusb/blob/master/src/class/net/net_device.c#L307) and [L.404](https://github.com/hathach/tinyusb/blob/master/src/class/net/net_device.c#L404). But, it seems that [L.404](https://github.com/hathach/tinyusb/blob/master/src/class/net/net_device.c#L404) only runs, in my Windows 10 host's log. If Linux hosts execute both lines, I guess the point has something hints for the issue.

**Reply by hathach** [reply]
@Wini-Buh I also encouter the ASSERT with 0x81 endpoint, however, it seems to be "**normal**" since currently the net driver doesn't check the busy state of the notification endpoint and can send one while previous one doesn't received by host yet. This will be fixed later but in general it does affect the example as this also appear on other working MCUs as well.
